lib/class-wp-rest-pattern-directory-controller.php

```diff
@@ -0,0 +1,262 @@
+		);
+	}
+
+	/**
+	 * Search and retrieve block patterns metadata
+	 *
+	 * @since 5.8.0
+	 *
+	 * @param WP_REST_Request $request Full details about the request.
+	 *
+	 * @return WP_Error|WP_REST_Response Response object on success, or WP_Error object on failure.
+	 */
+	public function get_items( $request ) {
+		$query_args   = array();
+		$category_ids = $request->get_param( 'category' );
+		$search_term  = $request->get_param( 'search' );
+
+		if ( $category_ids ) {
+			$query_args['pattern-categories'] = $category_ids;
+		}
+
+		if ( $search_term ) {
+			$query_args['search'] = $search_term;
+		}
+
+		$api_url = add_query_arg(
+			$query_args,
+			'http://api.wordpress.org/patterns/1.0/'
+		);
+
+		if ( wp_http_supports( array( 'ssl' ) ) ) {
+			$api_url = set_url_scheme( $api_url, 'https' );
+		}
+
+		$response     = wp_remote_get( $api_url );
+		$raw_patterns = json_decode( wp_remote_retrieve_body( $response ) );
+
+		if ( is_wp_error( $response ) ) {
+			$response->add_data( array( 'status' => 500 ) );
+
+			return $response;
+		}
+
+		$prepared_patterns = array();
+
+		if ( $raw_patterns ) {
+			foreach ( $raw_patterns as $pattern ) {
+				$prepared_patterns[] = $this->prepare_item_for_response( $pattern, $request );
+			}
+		}
+
+		return rest_ensure_response( $prepared_patterns );
+	}
+
+	/**
+	 * Prepare a raw pattern before it's output in an API response.
+	 *
+	 * @since 5.8.0
+	 *
+	 * @param object          $raw_pattern A pattern from api.wordpress.org, before any changes.
+	 * @param WP_REST_Request $request     Request object.
+	 *
+	 * @return array
+	 */
+	public function prepare_item_for_response( $raw_pattern, $request ) { // phpcs:ignore VariableAnalysis.CodeAnalysis.VariableAnalysis.UnusedVariable -- Method must match signature of parent class.
+		$prepared_pattern = array(
+			'id'             => absint( $raw_pattern->id ),
+			'title'          => sanitize_text_field( $raw_pattern->title->rendered ),
+			'content'        => wp_kses_post( $raw_pattern->content->rendered ),
+			'categories'     => array_map( 'sanitize_title', $raw_pattern->category_slugs ),
+			'keywords'       => array_map( 'sanitize_title', $raw_pattern->keyword_slugs ),
+			'description'    => sanitize_text_field( $raw_pattern->meta->wpop_description ),
+			'viewport_width' => absint( $raw_pattern->meta->wpop_viewport_width ),
+		);
+
+		return $prepared_pattern;
+	}
+
+	/**
+	 * Retrieves the pattern's schema, conforming to JSON Schema.
+	 *
+	 * @since 5.8.0
+	 *
+	 * @return array Item schema data.
+	 */
+	public function get_item_schema() {
+		if ( $this->schema ) {
+			return $this->add_additional_fields_schema( $this->schema );
+		}
+
+		$this->schema = array(
+			'$schema'    => 'http://json-schema.org/draft-04/schema#',
+			'title'      => 'pattern-directory-item',
+			'type'       => 'object',
+
+			'properties' => array(
+				'id'             => array(
+					'description' => __( 'The pattern ID.', 'gutenberg' ),
+					'required'    => true,
```

Comment:
We should drop all these `required` declarations, they are used for documenting that a request parameter is `required`.

Reply:
I added them because they're used in `rest_validate_value_from_schema()` (called from `assertPatternMatchesSchema()`). Is that not correct?

Reply:
I'd prefer manually adding them in the test so the published schema stays correct. `required` represents the request parameter is `required`, it doesn't say anything about the parameter being present in the response.

Reply:
Ah, I see. Done in db207d157b